torchaudio/csrc/ray_tracing.cpp

```diff
@@ -0,0 +1,577 @@
+      }
+    } else if (D == 2) {
+      scalar_t offset = 2. * M_PI / num_rays;
+      for (int i = 0; i < num_rays; ++i) {
+        simul_ray(histograms, i * offset, 0.);
+      }
+    }
+  }
+
+ private:
+  const torch::Tensor& room;
+  const torch::Tensor& source;
+  const torch::Tensor& mic_array;
+  int num_rays;
+  scalar_t energy_0; // initial energy of each ray
+  scalar_t mic_radius;
+  double mic_radius_sq;
+  scalar_t sound_speed;
+  scalar_t energy_thres;
+  scalar_t time_thres;
+  scalar_t hist_bin_size;
+  scalar_t max_dist; // Max distance needed to hit a wall = diagonal of room + 1
+  int D; // Dimension of the room
+  const bool do_scattering; // Whether scattering is needed (scattering != 0)
+  std::vector<Wall<scalar_t>> walls; // The walls of the room
+
+  /**
+   * From a ray vector defined by its start and end, returns the next wall hit
+   * as a 3-tuple:
+   * - the hit point on the wall: 2D or 3D tensor
+   * - the index of the wall (as in the .walls vector attribute)
+   * - the distance from the start to the wall
+   */
+  std::tuple<torch::Tensor, int, scalar_t> next_wall_hit(
+      const torch::Tensor& start,
+      const torch::Tensor& end) {
+    const static std::vector<std::vector<int>> shoebox_orders = {
+        {0, 1, 2}, {1, 0, 2}, {2, 0, 1}};
+
+    torch::Tensor hit_point = torch::zeros_like(room);
+    int next_wall_index = -1;
+    auto hit_dist = max_dist;
+
+    torch::Tensor dir = end - start;
+
+    auto dir_a = dir.accessor<scalar_t, 1>();
+    auto hit_point_a = hit_point.accessor<scalar_t, 1>();
+    auto start_a = start.accessor<scalar_t, 1>();
+    auto room_a = room.accessor<scalar_t, 1>();
+
+    for (auto& d : shoebox_orders) {
+      if (d[0] >= D) { // Happens for 2D rooms
+        continue;
+      }
+      auto abs_dir0 = std::abs(dir_a[d[0]]);
+      if (abs_dir0 < EPS) {
+        continue;
+      }
+
+      // distance to plane
+      auto distance = 0.;
+
+      // this will tell us if the front or back plane is hit
+      int ind_inc = 0;
+
+      if (dir_a[d[0]] < 0.) {
+        hit_point[d[0]] = 0.;
+        distance = start_a[d[0]];
+        ind_inc = 0;
+      } else {
+        hit_point_a[d[0]] = room_a[d[0]];
+        distance = room_a[d[0]] - start_a[d[0]];
+        ind_inc = 1;
+      }
+
+      if (distance < EPS) {
+        continue;
+      }
+
+      auto ratio = distance / abs_dir0;
+
+      // Now compute the intersection point and verify if intersection happens
+      auto intersection_found = true;
+      for (unsigned int i = 1; i < D; ++i) {
+        hit_point_a[d[i]] = start_a[d[i]] + ratio * dir_a[d[i]];
+        // when there is no intersection, we jump to the next plane
+        if ((hit_point_a[d[i]] <= -EPS) ||
+            (room_a[d[i]] + EPS <= hit_point_a[d[i]])) {
+          intersection_found = false;
+          break; // check next plane
+        }
+      }
+
+      if (intersection_found) {
+        next_wall_index = 2 * d[0] + ind_inc;
```

Comment:
@nateanl I feel like this should be `d[0] + 2 * ind_inc`. Do we have test to verify this is correct?

Reply:
@nateanl Actually, this is correct for 3D but not for 2D. Because for 2D the order of wall is SENW, while for 3D it's WESNCF.